omnibus/files/private-chef-ctl-commands/chef-server-ctl

```diff
@@ -0,0 +1,328 @@
+#!/opt/opscode/embedded/bin/ruby
+
+require 'rubygems'
+# Because we're not installed as a gem ourselves,
+# we'll need to explicitly load the omnibus-ctl gem
+# so that we can see it
+# so that we can see it.
+gem 'omnibus-ctl'
+require "omnibus-ctl"
+
+module Omnibus
+
+  # We keep the base behavior of Omnibus::Ctl, but
+  # add our own hooks for before/after command run.
+  # In particular we hook the service calls.
+  class ChefServerCtl < Omnibus::Ctl
+    def command_post_hook(*args)
+      command = args.shift
+      method = to_method_name("#{command}_post_hook")
+      if respond_to?(method)
+        send(method, *args)
+      end
+    end
+
+    # If a command has a pre-hook defined we will run it.
+    # Otherwise, if it is a run-sv command and the service it refers to
+    # is an external service, we will show an error since we
+    # can't control external services from here.
+    def command_pre_hook(*args)
+      command = args.shift
+      method = to_method_name("#{command}_pre_hook")
+      if respond_to?(method)
+        send(method, *args)
+      else
+        return true if args.empty?
+        # SV_COMMAND_NAMES is predefined in Omnibus::Ctl
+        # and contains a list of commands that we hand off to
+        # runsv to manage.
+        if SV_COMMAND_NAMES.include? command
+          if service_external? args[0]
+            log error_external_service(command, args[0])
+            return false;
+          end
+        end
+        true
+      end
+    end
+
+    # If we're listing status for all services and have external
+    # services to show, we'll include an output header to show that
+    # we're reporting internal services
+    def status_pre_hook(service = nil)
+      log_internal_service_header if service.nil?
+      true
+    end
+
+    # Status gets its own hook because each externalized service will
+    # have its own things to do in order to report status.
+    # As above, we may also include an output header to show that we're
+    # reporting on external services.
+    def status_post_hook(service = nil)
+      if service.nil?
+        log_external_service_header
+        external_services.each_key do |service_name|
+          status = send(to_method_name("external_status_#{service_name}"), :sparse)
+          log status
+        end
+      else
+        # Request verbose status if the service is asked for by name.
+        if service_external?(service)
+          status = send(to_method_name("external_status_#{service}"), :verbose)
+          log status
+        end
+      end
+    end
+
+    # Data cleanup requirements for external services aren't met by the standard
+    # 'nuke /var/opt' behavior - this hook allows each service to perform its own
+    # 'cleanse' operations.
+    def cleanse_post_hook(*args)
+        external_services.each_key do |service_name|
+          perform_delete = ARGV.include?("--with-remote")
```

Comment:
Minor point, but you can move this outside of the loop to avoid doing the look up on every iteration.